tuner/tuner/common_test.py

```diff
@@ -206,3 +206,120 @@ def test_get_lowering_config(tuner_ctx: common.TunerContext) -> None:
+
+
+def test_link_tuning_specs(tuner_ctx: common.TunerContext) -> None:
+    context = tuner_ctx.mlir_ctx
+    module_str = """
+        module @inner_module_a
+            attributes { transform.with_named_sequence, iree_codegen.tuning_spec_with_default_entrypoint } {
+            transform.named_sequence @match(%arg: !transform.any_op {transform.readonly}) -> (!transform.any_op) {
+                transform.yield %arg : !transform.any_op
+            }
+
+            transform.named_sequence @apply_op_config(%op: !transform.any_op {transform.readonly}) {
+                transform.yield
+            }
+
+            transform.named_sequence @__kernel_config(%arg0: !transform.any_op {transform.consumed})
+            -> (!transform.any_op) attributes { iree_codegen.tuning_spec_entrypoint } {
+                %res = transform.foreach_match in %arg0 @match -> @apply_op_config
+                    : (!transform.any_op) -> (!transform.any_op)
+                transform.yield %res : !transform.any_op
+            }
+        }
+    """
+
+    first_ir_module = ir.Module.parse(module_str, context)
+    second_ir_module = ir.Module.parse(module_str, context)
+    second_ir_module.operation.attributes["sym_name"] = ir.StringAttr.get(
+        "inner_module_b"
+    )
+    linked_module = common.link_tuning_specs(
+        tuner_ctx, [first_ir_module, second_ir_module]
+    )
+    assert linked_module
+
+    assert "transform.with_named_sequence" in linked_module.operation.attributes
+    assert (
+        "iree_codegen.tuning_spec_with_default_entrypoint"
+        in linked_module.operation.attributes
+    )
+
+    inner_ops = list(linked_module.body.operations)
+    # Check that inner modules have been merged into the top-level module and no inner modules remain.
+    assert all(
+        op.name != "builtin.module" for op in inner_ops
+    ), "Unexpected inner builtin.module ops found"
+
+    named_sequences = []
+    kernel_config_op = None
+    for op in linked_module.body.operations:
+        if op.name == "transform.named_sequence":
+            sym_name_attr = op.sym_name
+            assert sym_name_attr is not None
+            named_sequences.append(sym_name_attr.value)
+            if sym_name_attr.value == "__kernel_config":
+                kernel_config_op = op
+
+    assert kernel_config_op is not None, "Missing @__kernel_config"
+
+
+def test_link_tuning_specs_raises_error(tuner_ctx: common.TunerContext) -> None:
+    context = tuner_ctx.mlir_ctx
+    module_str = """
+        module @inner_module_a
+            attributes { transform.with_named_sequence } {
+            transform.named_sequence @match(%arg: !transform.any_op {transform.readonly}) -> (!transform.any_op) {
+                transform.yield %arg : !transform.any_op
+            }
+
+            transform.named_sequence @apply_op_config(%op: !transform.any_op {transform.readonly}) {
+                transform.yield
+            }
```

Comment:
Do we need these two named sequences to make it fail because of missing __kernel_config?